Simplify IterDomain extents if possible

naoyam opened the pull request:
Simplify extents of `Resize` output IterDomains if possible. Each resize output domain has an extent of `in->extent() + left_expansion + right_expansion`. In the case of `slice`, with `start` and `stop` offsets, the expansion parameters are `-start_offset` and `stop_offset - in->extent()`, respectively, so the overall extent is `in->extent() - start_offset + (end_offset - in->extent())`. Of course, this can be simplified to `end_offset - start_offset`.

Unexpectedly, this simplification exposed a problem with `ValReplacementMutator`, which is used by `replaceSymbolicSizes`. The underlying root cause is that IterVisitor does not traverse into attributes, such as the expansion parameters of `Resize`. It's fixed in #50.

This also fixes an error encountered with a recently added Python test, `test_nanogpt_split_mha_linears`. A C++ version of the test is added.

naoyam commented on csrc/ir_utils.h:
```diff
@@ -114,6 +114,14 @@ class FilteredView {
     return std::vector<value_type>(begin(), end());
   }
 
+  size_t size() const {
```
Ended up not using this, but think it's generally useful

naoyam commented on csrc/lower_utils.h:
```diff
@@ -103,6 +103,8 @@ c10::optional<IterDomain*> getMaybeWarpReductionDim(
 
 bool isScalarOp(const Expr*);
 
+bool isIterDomainOp(const Expr*);
```
Ended up not using this, but think it's generally useful

zasdfgbnm reviewed (approved):
This PR LGTM, but what is the failing test and why simplifying expr fixes it? I am OK with simplifying expr here because it has better readability, but I am not happy with the fact that correctness depends on simplification. Maybe the pass that generates this failure needs a `simplifyExpr` call?

naoyam commented:
Forgot to mention about the error. So, before this, the added C++ test failed as below:

```
C++ exception with description "Tried to bind to a value that is computed in the Fusion IR: ( T1.size[0] + ( 16 - T1.size[0] ) ) with 16
```

This is because the fusion is segmented, and one of the segments has an input like this:

```
T1_g[ iS77{( ceilDiv(( ceilDiv(( 16 * ( ceilDiv(( i3 + ( 1024 - i3 ) ), 16) ) ), 4) ), blockDim.x) )}, iS80{( ceilDiv(( ( i0 + ( 16 - i0 ) ) * ( i2 + ( 128 - i2 ) ) ), 1) )}, iS81{1}, iS76{4}, iS78{blockDim.x} ]
 root domain : (iS3{i0}rf,iS5{i2}rf,iS7{i3}rf)
  Resize: iS3{i0}rf by 0 and ( 16 - i0 ) -> iS4{( i0 + ( 16 - i0 ) )}rf
  Resize: iS5{i2}rf by 0 and ( 128 - i2 ) -> iS6{( i2 + ( 128 - i2 ) )}rf
  Resize: iS7{i3}rf by 0 and ( 1024 - i3 ) -> iS8{( i3 + ( 1024 - i3 ) )}rf
 rfactor domain : (iS4{( i0 + ( 16 - i0 ) )}rf,iS6{( i2 + ( 128 - i2 ) )}rf,iS8{( i3 + ( 1024 - i3 ) )}rf)
 contiguity: t t t
  Outer split: iS8{( i3 + ( 1024 - i3 ) )}rf by factor 16 -> iS48{16}, iS49{( ceilDiv(( i3 + ( 1024 - i3 ) ), 16) )}, start offset: 0, stop offset: 0
  Merge: iS48{16} and iS49{( ceilDiv(( i3 + ( 1024 - i3 ) ), 16) )} -> iS74{( 16 * ( ceilDiv(( i3 + ( 1024 - i3 ) ), 16) ) )}
  Split: iS74{( 16 * ( ceilDiv(( i3 + ( 1024 - i3 ) ), 16) ) )} by factor 4 -> iS75{( ceilDiv(( 16 * ( ceilDiv(( i3 + ( 1024 - i3 ) ), 16) ) ), 4) )}, iS76{4}, start offset: 0, stop offset: 0
  Split: iS75{( ceilDiv(( 16 * ( ceilDiv(( i3 + ( 1024 - i3 ) ), 16) ) ), 4) )} by factor blockDim.x -> iS77{( ceilDiv(( ceilDiv(( 16 * ( ceilDiv(( i3 + ( 1024 - i3 ) ), 16) ) ), 4) ), blockDim.x) )}, iS78{blockDim.x}, start offset: 0,
stop offset: 0
  Merge: iS4{( i0 + ( 16 - i0 ) )}rf and iS6{( i2 + ( 128 - i2 ) )}rf -> iS79{( ( i0 + ( 16 - i0 ) ) * ( i2 + ( 128 - i2 ) ) )}
  Split: iS79{( ( i0 + ( 16 - i0 ) ) * ( i2 + ( 128 - i2 ) ) )} by factor 1 -> iS80{( ceilDiv(( ( i0 + ( 16 - i0 ) ) * ( i2 + ( 128 - i2 ) ) ), 1) )}, iS81{1}, start offset: 0, stop offset: 0
```

When launching this kernel, ExpressionEvaluator failed as it couldn't bind 16 to the extent of the first axis of the rfactor root domain, `(iS4{( i0 + ( 16 - i0 ) )}rf`. Obviously, an easy fix is to simplify `i0 + (16 - i0)`, so that's why I changed `IterDomain::resize` to do this. I considered using the expr simplifier, but it's also quite trivial to do locally in `IterDomain::resize`, so that's what I did.

This however exposed the issue addressed in #50. One of the segments looks like:

```
%kernel {
T8_l[ iblockIdx.x58{( ceilDiv(( ceilDiv(( 128 * 1024 ), 4) ), blockDim.x) )}, iblockIdx.y60{( ceilDiv(16, 1) )}, iUS61{1}, iS57{4}, ithreadIdx.x59{blockDim.x} ] ca_pos( 3 )
   = slice( T0_g[ iS79{( ceilDiv(( ceilDiv(( T0.size[1] * T0.size[2] ), 4) ), blockDim.x) )}, iS81{( ceilDiv(T0.size[0], 1) )}, iS82{1}, iS78{4}, iS80{blockDim.x} ], { {0, 16, 1} {0, 128, 1} {0, 1024, 1} } )
T1_g[ iblockIdx.x51{( ceilDiv(( ceilDiv(( 128 * 1024 ), 4) ), blockDim.x) )}, iblockIdx.y53{( ceilDiv(16, 1) )}, iUS54{1}, iV50{4}, ithreadIdx.x52{blockDim.x} ] ca_pos( 3 ) produce_pos( 3 )
   = T8_l[ iblockIdx.x58{( ceilDiv(( ceilDiv(( 128 * 1024 ), 4) ), blockDim.x) )}, iblockIdx.y60{( ceilDiv(16, 1) )}, iUS61{1}, iS57{4}, ithreadIdx.x59{blockDim.x} ] ca_pos( 3 );

TransformPrinter :
T0_g[ iS79{( ceilDiv(( ceilDiv(( T0.size[1] * T0.size[2] ), 4) ), blockDim.x) )}, iS81{( ceilDiv(T0.size[0], 1) )}, iS82{1}, iS78{4}, iS80{blockDim.x} ]
 root domain : (iS85{T0.size[0]},iS83{T0.size[1]},iS84{T0.size[2]})
 contiguity: f f f
  Merge: iS83{T0.size[1]} and iS84{T0.size[2]} -> iS76{( T0.size[1] * T0.size[2] )}
  Split: iS76{( T0.size[1] * T0.size[2] )} by factor 4 -> iS77{( ceilDiv(( T0.size[1] * T0.size[2] ), 4) )}, iS78{4}, start offset: 0, stop offset: 0
  Split: iS77{( ceilDiv(( T0.size[1] * T0.size[2] ), 4) )} by factor blockDim.x -> iS79{( ceilDiv(( ceilDiv(( T0.size[1] * T0.size[2] ), 4) ), blockDim.x) )}, iS80{blockDim.x}, start offset: 0, stop offset: 0
  Split: iS85{T0.size[0]} by factor 1 -> iS81{( ceilDiv(T0.size[0], 1) )}, iS82{1}, start offset: 0, stop offset: 0
T8_l[ iblockIdx.x58{( ceilDiv(( ceilDiv(( 128 * 1024 ), 4) ), blockDim.x) )}, iblockIdx.y60{( ceilDiv(16, 1) )}, iUS61{1}, iS57{4}, ithreadIdx.x59{blockDim.x} ] ca_pos( 3 )
 root domain : (iS88{T0.size[0]}rf,iS86{T0.size[1]}rf,iS87{T0.size[2]}rf)
  Resize: iS88{T0.size[0]}rf by 0 and ( 16 - i0 ) -> iS4{16}rf
  Resize: iS86{T0.size[1]}rf by 0 and ( 128 - i2 ) -> iS6{128}rf
  Resize: iS87{T0.size[2]}rf by 0 and ( 1024 - i3 ) -> iS8{1024}rf
 rfactor domain : (iS4{16}rf,iS6{128}rf,iS8{1024}rf)
 contiguity: t t t
  Merge: iS6{128}rf and iS8{1024}rf -> iS55{( 128 * 1024 )}
  Split: iS55{( 128 * 1024 )} by factor 4 -> iS56{( ceilDiv(( 128 * 1024 ), 4) )}, iS57{4}, start offset: 0, stop offset: 0
  Split: iS56{( ceilDiv(( 128 * 1024 ), 4) )} by factor blockDim.x -> iblockIdx.x58{( ceilDiv(( ceilDiv(( 128 * 1024 ), 4) ), blockDim.x) )}, ithreadIdx.x59{blockDim.x}, start offset: 0, stop offset: 0
  Split: iS4{16}rf by factor 1 -> iblockIdx.y60{( ceilDiv(16, 1) )}, iUS61{1}, start offset: 0, stop offset: 0
T1_g[ iblockIdx.x51{( ceilDiv(( ceilDiv(( 128 * 1024 ), 4) ), blockDim.x) )}, iblockIdx.y53{( ceilDiv(16, 1) )}, iUS54{1}, iV50{4}, ithreadIdx.x52{blockDim.x} ] ca_pos( 3 ) produce_pos( 3 )
 root domain : (iS42{16},iS43{128},iS44{1024})
 contiguity: t t t
  Merge: iS43{128} and iS44{1024} -> iS48{( 128 * 1024 )}
  Split: iS48{( 128 * 1024 )} by factor 4 -> iS49{( ceilDiv(( 128 * 1024 ), 4) )}, iV50{4}, start offset: 0, stop offset: 0
  Split: iS49{( ceilDiv(( 128 * 1024 ), 4) )} by factor blockDim.x -> iblockIdx.x51{( ceilDiv(( ceilDiv(( 128 * 1024 ), 4) ), blockDim.x) )}, ithreadIdx.x52{blockDim.x}, start offset: 0, stop offset: 0
  Split: iS42{16} by factor 1 -> iblockIdx.y53{( ceilDiv(16, 1) )}, iUS54{1}, start offset: 0, stop offset: 0
}
```

The issue is `T8`. While its rfactor domain just has static-sized axes, because of the simplification, its root  and the resize ops from the root to the rfactor have `i0`, `i2` and `i3`, which are equivalent to `T0.size[0]`, `T0.size[1]` and `T0.size[1]`. Note that the above printing of the fusion IR is done after `replaceSymbolicSizes`, so these symbols should have been replaced with the actual sizes, but that didn't happen due to the issue of #50. Because of that, the generated kernel still had those symbols, and the  nvrtc compilation failed.

This issue of #50 existed before the extent simplification but just wasn't exposed. Without the simplification, the output domain of a `Resize` op has an extent of `input_domain->extent() + left_expansion + right_expansion`. For example, the first RF axis of `T8` would have been `i0 + 0 + (16 - i0)`. And `ValReplacementMutator` replaces each term, `i0` with `T0.size[0]` and `(16 - i0)` with `(16 - T0.size[0])`, and this replacement also affects the resize op, `Resize: iS88{T0.size[0]}rf by 0 and ( 16 - i0 ) -> iS4{16}rf`, as its expansion attributes exactly use the same `Val`s as those used as part of the output extent, so the resize op becomes: `Resize: iS88{T0.size[0]}rf by 0 and ( 16 - T0.size[0] ) -> iS4{16}rf``. 

When the simplification is done, the output extent just becomes `(16)`, so the expansion attributes remain to have the `i0` symbol. So, it worked before, but it was (at least to me) unintentional and just by luck.

So, overall, the above issue was fixed by doing the simplification and extending IterVisitor to traverse into attributes. However, these fixes may not be sufficient in general. With the Resize op, we could still end up having an IterDomain with extent like `T1.size[0] + a`, e.g, with padding. If such extents showed up as an axis of an input, ExpressionEvaluator would still fail. I'd expect several iterations of fixes/extensions still needed to iron out  the Resize op.

zasdfgbnm commented:
Thanks for the detailed explanation!
